Fix size and empty methods in relations with group and order

jmbejar opened the pull request:
Fix the problem mentioned in this issue https://github.com/rails/rails/issues/20508

The idea is to prevent invalid SQL count queries created on the `empty?` and `size` methods by excluding the order clause. The issue is happening when the order clause depends on an expression using the SQL count function because this part is changed by AR at the moment of creating the SQL count query.
After all, removing the order clause in a SQL count query does not alter the result.

rafaelfranca commented:
> 'm aware these changes are not backward compatible, however the current behavior of size in relations with a group value is not consistent, so I think it should be fixed to make sure that it always returns the number of groups (like it does if it's called in a loaded relation).

It is documented this way. If you want consistent behavior use `count` instead of `size` to always get the value from the database.

rafaelfranca commented:
I'll test the impact of each change individually before rejecting

jmbejar commented:
@rafaelfranca Could you point me to the documentation where `size` behavior is explained? I would help me to better understand the implications of this change if the behavior is actually documented (I couldn't find by myself anything related in guides or api docs).

IMHO, I don't see strong reasons to have such inconsistency in this method. In fact, it looks like an implementation mistake in the `size` method:

```ruby
     # Returns size of the records.
     def size
          loaded? ? @records.length : count(:all)`
     end
```

It seems that this implementation is not taking into consideration how `count` behaves with and without a `group` clause, returning a hash in the latter case. However, `@records.length` will always return an integer, even if the relation has a `group` statement. For this reason, when the relation is already loaded the **type of the result** ends up being different.

Said that, I found this inconsistency when working on fixing #20508. I think it leaded to a cleaner solution, however I will be happy on helping to solve this issue without changing `size` behavior if we arrive to that conclusion.

jmbejar committed:
Fix Relation#empty and #any? method to do not fail when an order clause
is present.

It helps to prevent invalid queries when the order clause
depends on expresions using the SQL count that are replaced by the
`count(:all)` logic in AR.

The fix is to exclude the order clause from the count query
(the order clause does not alter the result of the count anyway)

Solves #20508.

jmbejar committed:
Fix relation#size method for cases with group and order clauses

Order clauses can be excluded when forming the SQL query for the count,
since it does not alter the result. This change prevents issues like
happens with the #size method)

jmbejar commented:
@rafaelfranca I was thinking again on this issue. I understand now the implications of the change I made on the `#size` method, you're right about the way it is documented.

I just updated this PR description and added code that just fix the issue #20508 (and it also fix the same problem in the `#size` method given it has a very similar implementation).

Sorry for the initial confusion trying to solve too many things in the PR. Hopefully, it should be easier to review now.

fabianrbz commented:
@rafaelfranca @jmbejar I just came across this issue in the app I'm currently working on.
What's the status of this? Is there anything I can help with?

jmbejar commented:
I'm just rebased this branch and tests still pass, so I think it's ready to merge if we want to do it now.

@rafaelfranca Let me know if I can do anything else to help with this issue.

kamipo commented:
Closing this since #20508 has been fixed.
